### DISCUSSION TITLE
Bankable-Smart-Control-Change-Potentiometer

### DISCUSSION THREAD

**Discussion by milkos86** [tttapa/Control-Surface#631] [Q&A]
Hi tell me if Bankable-Smart-Control-Change-Potentiometer can be controlled on multiplexor (s) 74HC4051
and if so is there an example?

**Comment by tttapa** [comment]
Yes. The principle for “smart” potentiometers is exactly the same as shown in the [Getting Started guide](https://tttapa.github.io/Control-Surface-doc/Doxygen/d5/d7d/md_pages_Getting-Started.html).

**Reply by milkos86** [reply]
I am getting an error like this, what did I do wrong?
sketch_dec12a:33:1: error: no matching function for call to 'CS::Bankable::CCSmartPotentiometer<4>::CCSmartPotentiometer(<brace-enclosed initializer list>)'
 };
 
```cpp
#include <Control_Surface.h> // Include the Control Surface library

// Instantiate a MIDI over USB interface.
USBDebugMIDI_Interface midi;

// Four banks to change the channel of the potentiometer
Bank<4> bank;

// Selector to change banks
IncrementDecrementSelector<4> selector {bank, {2, 3}};

// Handy type alias
using  CCSmartPot = Bankable::CCSmartPotentiometer<4>;

// Instantiate an analog multiplexer
CD74HC4051 mux {
  A0,       // Analog input pin
  {3, 4, 5} // Address pins S0, S1, S2
};

// Instantiate a CCPotentiometer object
CCSmartPot potentiometer {
  {bank, BankType::CHANGE_CHANNEL},     // Bank configuration
  A0,                                   // Analog pin connected to potentiometer
  {mux.pin(0), {MIDI_CC::Channel_Volume, CHANNEL_1}},
  {mux.pin(1), {MIDI_CC::Channel_Volume, CHANNEL_2}},
  {mux.pin(2), {MIDI_CC::Channel_Volume, CHANNEL_3}},
  {mux.pin(3), {MIDI_CC::Channel_Volume, CHANNEL_4}},
  {mux.pin(4), {MIDI_CC::Channel_Volume, CHANNEL_5}},
  {mux.pin(5), {MIDI_CC::Channel_Volume, CHANNEL_6}},
  {mux.pin(6), {MIDI_CC::Channel_Volume, CHANNEL_7}},
  {mux.pin(7), {MIDI_CC::Channel_Volume, CHANNEL_8}},
};

void setup() {
  Control_Surface.begin(); // Initialize Control Surface
}

void loop() {
  Control_Surface.loop(); // Update the Control Surface

  // You can request the state of the smart potentiometer:
  if (potentiometer.getState() == CCSmartPot::Lower) {
    // Turn on an LED to indicate that the potentiometer
    // value is too low to be active
  } else if (potentiometer.getState() == CCSmartPot::Higher) {
    // Turn on an LED to indicate that the potentiometer 
    // value is too high to be active
  }
}
```

**Reply by tttapa** [reply]
> ```c++
> // Instantiate a CCPotentiometer object
> CCSmartPot potentiometer {
>   {bank, BankType::CHANGE_CHANNEL},     // Bank configuration
>   A0,                                   // Analog pin connected to potentiometer
>   {mux.pin(0), {MIDI_CC::Channel_Volume, CHANNEL_1}},
>   ...
> ```

This is not the correct way of initializing an instance of `CCSmartPotentiometer`, please see the documentation for its constructor and look at the arguments: https://tttapa.github.io/Control-Surface-doc/Doxygen/d3/dfe/classBankable_1_1CCSmartPotentiometer.html#ab12e9a0662c2f053926b95e45934d409

For example:
```cpp
CCSmartPot potentiometer {
  {bank, BankType::CHANGE_CHANNEL},     // config
  mux.pin(0),                           // analogPin
  {MIDI_CC::Channel_Volume, CHANNEL_1}, // address
};
```

**Reply by milkos86** [reply]
```cpp
#include <Control_Surface.h> // Include the Control Surface library

// Instantiate a MIDI over USB interface.
USBDebugMIDI_Interface midi;

// Four banks to change the channel of the potentiometer
//Bank<7> bank;

// Selector to change banks


// Создаем 4 банка, по 2 трека на банк 
Bank < 7 > bank ( 2 );

// Создаем новый селектор банка, который изменяет настройку банка только что созданного 
// Он имеет кнопки, подключенные к контактам 10 и 11, которые увеличивают или уменьшают настройку банка, 
// и 4 светодиода к контактам 5, 6, 7 и 8, отображающие текущую настройку банка. 
IncrementDecrementSelectorLEDs < 7 > selector = {
  bank,          // банк для управления 
  { 2, 3 },      // выводы кнопок 
  { 4, 5, 6, 7, 11, 12, 13 }, // выводы светодиода 
  Wrap :: Wrap,    // переход на максимальное / минимальное значение (альтернатива = Wrap :: Зажим)
};

//IncrementDecrementSelectorLEDs<7> selector = {
//  bank,         // bank to control
//  {2, 3},     // button pins
//  {4, 5, 6, 7, 11, 12, 13}, // LED pins
//  Wrap::Wrap,   // wrap around at the maximum/minimum setting (alternative = Wrap::Clamp)
//};

// Handy type alias
using CCSmartPot = Bankable::CCSmartPotentiometer<7>;

// Instantiate an analog multiplexer
CD74HC4051 mux { A0, {8, 9, 10}}; // Address pins S0, S1, S2
CD74HC4051 mux1 { A1, {8, 9, 10}}; // Address pins S0, S1, S2

// Instantiate a CCPotentiometer object
CCSmartPot potentiometer0 {
  {bank, BankType::CHANGE_CHANNEL},
  mux.pin(0),
  {MIDI_CC::Channel_Volume, CHANNEL_1}
};

CCSmartPot potentiometer1 {
  {bank, BankType::CHANGE_CHANNEL},
  mux.pin(1),
  {MIDI_CC::Channel_Volume, CHANNEL_2}
};
CCSmartPot potentiometer2 {
  {bank, BankType::CHANGE_CHANNEL},
  mux.pin(2),
  {MIDI_CC::Channel_Volume, CHANNEL_3}
};
CCSmartPot potentiometer3 {
  {bank, BankType::CHANGE_CHANNEL},
  mux.pin(3),
  {MIDI_CC::Channel_Volume, CHANNEL_4}
};
CCSmartPot potentiometer4 {
  {bank, BankType::CHANGE_CHANNEL},
  mux.pin(4),
  {MIDI_CC::Channel_Volume, CHANNEL_5}
};
CCSmartPot potentiometer5 {
  {bank, BankType::CHANGE_CHANNEL},
  mux.pin(5),
  {MIDI_CC::Channel_Volume, CHANNEL_6}
};
CCSmartPot potentiometer6 {
  {bank, BankType::CHANGE_CHANNEL},
  mux.pin(6),
  {MIDI_CC::Channel_Volume, CHANNEL_7}
};
CCSmartPot potentiometer7 {
  {bank, BankType::CHANGE_CHANNEL},
  mux.pin(7),
  {MIDI_CC::Channel_Volume, CHANNEL_8}
};
CCSmartPot potentiometer8 {
  {bank, BankType::CHANGE_CHANNEL},
  mux1.pin(8),
  {MIDI_CC::Channel_Volume, CHANNEL_9}
};
CCSmartPot potentiometer9 {
  {bank, BankType::CHANGE_CHANNEL},
  mux1.pin(9),
  {MIDI_CC::Channel_Volume, CHANNEL_10}
};
CCSmartPot potentiometer10 {
  {bank, BankType::CHANGE_CHANNEL},
  mux1.pin(10),
  {MIDI_CC::Channel_Volume, CHANNEL_11}
};
CCSmartPot potentiometer11 {
  {bank, BankType::CHANGE_CHANNEL},
  mux1.pin(11),
  {MIDI_CC::Channel_Volume, CHANNEL_12}
};
CCSmartPot potentiometer12 {
  {bank, BankType::CHANGE_CHANNEL},
  mux1.pin(12),
  {MIDI_CC::Channel_Volume, CHANNEL_13}
};
CCSmartPot potentiometer13 {
  {bank, BankType::CHANGE_CHANNEL},
  mux1.pin(13),
  {MIDI_CC::Channel_Volume, CHANNEL_14}
};
CCSmartPot potentiometer14 {
  {bank, BankType::CHANGE_CHANNEL},
  mux1.pin(14),
  {MIDI_CC::Channel_Volume, CHANNEL_15}
};
CCSmartPot potentiometer15 {
  {bank, BankType::CHANGE_CHANNEL},
  mux1.pin(15),
  {MIDI_CC::Channel_Volume, CHANNEL_16}
};
void setup() {
  Control_Surface.begin(); // Initialize Control Surface
}

void loop() {
  Control_Surface.loop(); // Update the Control Surface
  // You can request the state of the smart potentiometer:
  if (potentiometer.getState() == CCSmartPot::Lower) {
    // Turn on an LED to indicate that the potentiometer
    // value is too low to be active
  } else if (potentiometer.getState() == CCSmartPot::Higher) {
    // Turn on an LED to indicate that the potentiometer 
    // value is too high to be active
  }
}
```

Hi, I'm not really into the code, I was advised to do this. But that doesn't work at all either. They said to remove this code. 
```cpp
// You can request the state of the smart potentiometer:
  if (potentiometer.getState () == CCSmartPot :: Lower) {
    // Turn on an LED to indicate that the potentiometer
    // value is too low to be active
  } else if (potentiometer.getState () == CCSmartPot :: Higher) {
    // Turn on an LED to indicate that the potentiometer
    // value is too high to be active
  }
```
The sketch loaded, but the potentiometers did not respond at all, When I returned it, an error began to appear
```
Arduino: 1.8.13 (Windows 10), Board: "Arduino Uno"
D: \ MRPP \ MIDI BankBle \ Midi_banckBle \ Midi_banckBle.ino: In function 'void loop ()':
Midi_banckBle: 128: 7: error: 'potentiometer' was not declared in this scope
   if (potentiometer.getState () == CCSmartPot :: Lower) {
       ^ ~~~~~~~~~~~~
D: \ MRPP \ MIDI BankBle \ Midi_banckBle \ Midi_banckBle.ino: 128: 7: note: suggested alternative: 'potentiometer9'
   if (potentiometer.getState () == CCSmartPot :: Lower) {
       ^ ~~~~~~~~~~~~
       potentiometer9
exit status 1
'potentiometer' was not declared in this scope
This report will have more information from
enabled option File -> Settings ->
Show verbose output at compile time
```

**Reply by tttapa** [reply]
> The sketch loaded, but the potentiometers did not respond at all
> ```cpp
>   mux1.pin(8),
> ```

Each multiplexer has only 8 pins, `mux1.pin(8)` does not exist.

> When I returned it, an error began to appear

The compiler is right, you never declare a variable with the name `potentiometer`, you only have variables with names `potentiometer0`, `potentiometer1`, etc.  
If you have variable names that end with a number, you should probably look into arrays and `for` loops.

**Reply by milkos86** [reply]
Is there any solution to this issue?
I don't understand what to do next. I seem to have followed all your instructions. Once again, I don’t understand how object oriented programming works. Putting together the code intuitively. I am not able to process arrays and data.

**Reply by tttapa** [reply]
> Is there any solution to this issue?

Just pass valid pin numbers to the `mux.pin(...)` method, i.e. indices from 0 to 7 (inclusive).

>  Once again, I don’t understand how object oriented programming works. [...] I am not able to process arrays and data.

If you want to use the code that generates errors to turn on/off LEDs, you'll have to learn how that works then :)

**Reply by milkos86** [reply]
Yes, I corrected this inattention right away. 
>Just pass valid pin numbers to the mux.pin(...) method, i.e. indices from 0 to 7 (inclusive).

But the code still doesn't work,
If I strip out the code at the end
     // You can request the state of the smart potentiometer:
   if (potentiometer.getState () == CCSmartPot :: Lower) {
     // Turn on an LED to indicate that the potentiometer
     // value is too low to be active
   } else if (potentiometer.getState () == CCSmartPot :: Higher) {
     // Turn on an LED to indicate that the potentiometer
     // value is too high to be active
   }
  
potentiometers, when scrolling the knob, switch channels and do not assign it.

#include <Control_Surface.h> // Include the Control Surface library

// Instantiate a MIDI over USB interface.
USBDebugMIDI_Interface midi;


// Создаем 4 банка, по 2 трека на банк 
Bank < 7 > bank;

// Создаем новый селектор банка, который изменяет настройку банка только что созданного 
// Он имеет кнопки, подключенные к контактам 10 и 11, которые увеличивают или уменьшают настройку банка, 
// и 4 светодиода к контактам 5, 6, 7 и 8, отображающие текущую настройку банка. 
IncrementDecrementSelectorLEDs < 7 > selector = {
  bank,          // банк для управления 
  { 2, 3 },      // выводы кнопок 
  { 4, 5, 6, 7, 11, 12, 13 }, // выводы светодиода 
  Wrap :: Wrap,    // переход на максимальное / минимальное значение (альтернатива = Wrap :: Зажим)
};

// Handy type alias
using CCSmartPot = Bankable::CCSmartPotentiometer<7>;

// Instantiate an analog multiplexer
CD74HC4051 mux { A0, {8, 9, 10}}; // Address pins S0, S1, S2
CD74HC4051 mux1 { A1, {8, 9, 10}}; // Address pins S0, S1, S2

// Instantiate a CCPotentiometer object
CCSmartPot potentiometer0 {
  {bank, BankType::CHANGE_CHANNEL},
  {mux.pin(0)},
  {MIDI_CC::Channel_Volume, CHANNEL_1},
};

CCSmartPot potentiometer1 {
  {bank, BankType::CHANGE_CHANNEL},
  {mux.pin(1)},
  {MIDI_CC::Channel_Volume, CHANNEL_2},
};
CCSmartPot potentiometer2 {
  {bank, BankType::CHANGE_CHANNEL},
  {mux.pin(2)},
  {MIDI_CC::Channel_Volume, CHANNEL_3},
};
CCSmartPot potentiometer3 {
  {bank, BankType::CHANGE_CHANNEL},
  {mux.pin(3)},
  {MIDI_CC::Channel_Volume, CHANNEL_4},
};
CCSmartPot potentiometer4 {
  {bank, BankType::CHANGE_CHANNEL},
  {mux.pin(4)},
  {MIDI_CC::Channel_Volume, CHANNEL_5},
};
CCSmartPot potentiometer5 {
  {bank, BankType::CHANGE_CHANNEL},
  {mux.pin(5)},
  {MIDI_CC::Channel_Volume, CHANNEL_6},
};
CCSmartPot potentiometer6 {
  {bank, BankType::CHANGE_CHANNEL},
  {mux.pin(6)},
  {MIDI_CC::Channel_Volume, CHANNEL_7},
};
CCSmartPot potentiometer7 {
  {bank, BankType::CHANGE_CHANNEL},
  {mux.pin(7)},
  {MIDI_CC::Channel_Volume, CHANNEL_8},
};
CCSmartPot potentiometer8 {
  {bank, BankType::CHANGE_CHANNEL},
  {mux1.pin(0)},
  {MIDI_CC::Channel_Volume, CHANNEL_9},
};
CCSmartPot potentiometer9 {
  {bank, BankType::CHANGE_CHANNEL},
  {mux1.pin(1)},
  {MIDI_CC::Channel_Volume, CHANNEL_10},
};
CCSmartPot potentiometer10 {
  {bank, BankType::CHANGE_CHANNEL},
  {mux1.pin(2)},
  {MIDI_CC::Channel_Volume, CHANNEL_11},
};
CCSmartPot potentiometer11 {
  {bank, BankType::CHANGE_CHANNEL},
  {mux1.pin(3)},
  {MIDI_CC::Channel_Volume, CHANNEL_12},
};
CCSmartPot potentiometer12 {
  {bank, BankType::CHANGE_CHANNEL},
  {mux1.pin(4)},
  {MIDI_CC::Channel_Volume, CHANNEL_13},
};
CCSmartPot potentiometer13 {
  {bank, BankType::CHANGE_CHANNEL},
  {mux1.pin(5)},
  {MIDI_CC::Channel_Volume, CHANNEL_14},
};
CCSmartPot potentiometer14 {
  {bank, BankType::CHANGE_CHANNEL},
  {mux1.pin(6)},
  {MIDI_CC::Channel_Volume, CHANNEL_15},
};
CCSmartPot potentiometer15 {
  {bank, BankType::CHANGE_CHANNEL},
  {mux1.pin(7)},
  {MIDI_CC::Channel_Volume, CHANNEL_16},
};
void setup() {
  Control_Surface.begin(); // Initialize Control Surface
}

void loop() {
  Control_Surface.loop(); // Update the Control Surface


}
